File: REST/k-means/src/main/java/local/example/outcome/OutcomeResource.java
package local.example.outcome;

import local.example.outcome.model.ClusterView;

import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Set;

@Path("/outcome")
public class OutcomeResource {

    private final Set<ClusterView> clusterViews = Collections.newSetFromMap(
        Collections.synchronizedMap(
                new LinkedHashMap<>()
        )
    );

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Response read() {
        if (clusterViews.isEmpty()) {
            return Response.ok().build();
        }
        return Response.ok(clusterViews).build();
    }

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Set<ClusterView> compute(ClusterView clusterView) {
        clusterView.setView();
        clusterViews.add(clusterView);
        return clusterViews;
    }
}

File: standard-deviation/src/main/kotlin/local/example/outcome/OutcomeResource.kt
package local.example.outcome

import local.example.outcome.model.Sigma

import java.util.*

import javax.ws.rs.GET
import javax.ws.rs.Path
import javax.ws.rs.Produces
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("/outcome")
class OutcomeResource {

    private val standardDeviations = Collections.newSetFromMap(
        Collections.synchronizedMap(
            LinkedHashMap<Sigma, Boolean>()
        )
    )

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun read(): Response? {
        return if (
            standardDeviations.isEmpty()
        ) Response.ok().build() else Response.ok(standardDeviations).build()
    }
}
